app/scripts/modules/core/utils/stickyHeader/Sticky.tsx

```diff
@@ -0,0 +1,169 @@
+
+    let width: number,
+        top: number,
+        zIndex = 5,
+        updatedParentPaddingTop = 0;
+
+    // If the element is supposed to be sticky, calculate the needed layout values
+    if (isSticky) {
+      const stickyElementRect = this.stickyElement.getBoundingClientRect();
+      const stickyElementWidth = stickyElementRect.width;
+      const stickyElementHeight = this.outerHeight(this.stickyElement);
+
+      updatedParentPaddingTop = stickyElementHeight;
+
+      if (containerTop + stickyElementHeight > sectionBottom) {
+        top = sectionBottom - stickyElementHeight;
+        zIndex = 4;
+      } else {
+        top = containerTop;
+      }
+
+      width = stickyElementWidth;
+    }
+
+    // Update the parent padding when it changed (either add padding
+    // for the height of the sticky element, or reset the height when
+    // the element is no longer sicky)
+    if (updatedParentPaddingTop !== this.sectionPaddingTop) {
+      this.sectionPaddingTop = updatedParentPaddingTop;
+      parentElement.style.paddingTop = `${this.sectionPaddingTop}px`;
+    }
+
+    if (isSticky !== this.state.isSticky ||
+        width !== this.state.width ||
+        top !== this.state.top ||
+        zIndex !== this.state.zIndex) {
+      this.setState({ width, top, isSticky, zIndex });
+    }
+  }
+
+  private addEventListeners(callback: () => void): void {
+    Sticky.RECOMPUTE_EVENTS.forEach((event) => {
+      this.stickyContainerElement.addEventListener(event, callback);
+    });
+    this.recomputeState();
+  }
+
+  private removeEventListeners(callback: () => void): void {
+    if (this.stickyContainerElement) {
+      Sticky.RECOMPUTE_EVENTS.forEach((event) => {
+        this.stickyContainerElement.removeEventListener(event, callback);
+      });
+    }
+  }
+
+  private refCallback = (element: HTMLElement): void => {
+    this.stickyElement = element;
+  }
+
+  public componentWillReceiveProps(nextProps: IProps): void {
+    if (nextProps.stickyIf !== this.props.stickyIf) {
+      if (nextProps.stickyIf) {
+        this.addEventListeners(this.recomputeState);
+      } else {
+        this.removeEventListeners(this.recomputeState);
+      }
+    }
+    this.setState({passthroughProps: this.passthroughProps(nextProps)});
+  }
+
+  public componentWillUnmount(): void {
+    this.removeEventListeners(this.recomputeState);
+  }
+
+  public render(): React.ReactElement<Sticky> {
+    let className = this.props.className || '';
+    if (this.state.isSticky) {
+      className += ' heading-sticky';
+    }
+    const style = this.props.style || {};
+    style.top = this.state.top;
+    style.width = this.state.width;
+    style.zIndex = this.state.zIndex;
+
+    return (
+      <div {...this.state.passthroughProps} className={className} style={style} ref={this.refCallback}>
+        {this.props.children}
+      </div>
+    );
+  }
```

Comment:
Much as I'd like to pretend to have reviewed this, I can barely bring myself to look at the existing sticky header code, so am assuming it's fine and you did a good job.